src/Clash/Cores/Ethernet/Examples/FullUdpStack.hs

```diff
@@ -0,0 +1,282 @@
+4. Finally, a `packetArbiterC` running in `RoundRobin` mode switches in a round-robin fashion between outputting a packet from the @arpEthOut stream and the @ipEthOut stream, i.e.
+   it alternatingly will output a complete packet coming from `toEthernetC` and a complete packet from `icmpUdpPacket`.
+
+All code thus far described handles interfacing between the MAC and IP layer. `icmpUdpStack` handles the payload of the IP packets, implementing two functionalities:
+ - sending an appropriate ICMP echo response to an ICMP echo request. 
+ - depacketizing, handling and packetizing UDP traffic with the provided UPD handler circuit.
+`icmpUdpStack`, line by line, does the following:
+
+1. The `packetDispatcher` + `routeBy` combination is now used to splite the input stream of type `PacketStream dom 4 IPv4HeaderLite` based on the protocol in the payload. The protocol number in the IP v4 header is also present iin `IPv4HeaderLite`, namely 
+   the field `_ipv4lProtocol`. 
+2. The stream `icmpIn` of ICMP packets is passed to `icmpEchoResponderC`. This circuit takes in a `Signal IPv4Adress` namely our IP, to internally
+  - parses the ICMP header from the content of the packet using the generic depacketizer and puts this into the metadata in a tuple with the 
+    already present `IPv4HeaderLite`
+  - updates simultaneously the IPv4HeaderLite and the IcmpHeader by 
+    - in the IP header, putting the source of the IP packet into the destination
+    - in the IP header, putting our own IP address into the destination of the ICMP packet.  
+    - in the ICMP header, updating the checksum according to the new type + code combination: we go from (8,0) (echo request) to (0,0) (echo response) and this requires a slight update of the checksum.
+   As said, the `icmpEchoResponderC` takes a signal of our IP, so we need to extract the `fst` of the input signal with `fst <$>`.
+3. The UDP stream is parsed using `udpDepacketizerC`, which returns a `PacketStream dom dataWidth (IPv4Address, UdpHeaderLite)`: it puts the parsed UDP header in a tuple: 
+   the `IPv4Address` in the tuple is the source address of the IPv4Packet. The UdpHeaderLite records the source port, destination port and payload length. All are of type `Unsigned 16`.
+4. The parsed UDP stream is then handled by the passed circuit.
+5. The handled UDP stream is then packetized again by `udpPacketizerC: this packetizer takes in a signal with our IP address and gives a circuit of type 
+   `Circuit (PacketStream dom n (IPv4Address, UdpHeaderLite)) (PacketStream dom n IPv4HeaderLite)`. It turns the `UdpHeaderLite` header into a proper `UdpHeader` by setting all 3 mandatory fields (source port, destination port, length)
+   with the fields provided by `UdpHeaderLite`, and puts @0x0000 in the checksum field of the UDP header. In IPv4, the checksum field for UDP is optional, and is ignored when set to @0x0000, so this will not cause trouble for the acceptance of the segments.
```

Comment:
your monospace is leaking!

Reply:
Yes, spaced monospace can be done by putting it between  @blabla@